// featherdown/dist/chunk-YKLQHPRB.js
import { visit, SKIP } from 'unist-util-visit';
import { toText } from 'hast-util-to-text';
import rehypeSanitize, { defaultSchema } from 'rehype-sanitize';
import rehypeAutolinkHeadings from 'rehype-autolink-headings';
import rehypeHighlight from 'rehype-highlight';
import rehypeKatex from 'rehype-katex';
import rehypeRaw from 'rehype-raw';
import rehypeSlug from 'rehype-slug';
import rehypeStringify from 'rehype-stringify';
import remarkDirective from 'remark-directive';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import { unified } from 'unified';

// src/internal/rehypeCodeBlockTitles.ts
var CHART_LANGUAGE_PREFIX = "language-chart-";
var WRAPPER_CLASS = "code-block";
var COPYABLE_CLASS = "code-block-copyable";
var COPY_BUTTON_CLASS = "code-block-copy-button";
function hasChartLanguage(className) {
  const values = Array.isArray(className) ? className : typeof className === "string" ? [className] : [];
  return values.some((value) => typeof value === "string" && value.startsWith(CHART_LANGUAGE_PREFIX));
}
var rehypeCodeBlockTitles = function rehypeCodeBlockTitles2() {
  return (tree) => {
    let copyTargetCount = 0;
    visit(tree, "element", (node, index, parent) => {
      if (node.tagName !== "pre") return;
      if (!parent || typeof index !== "number") return;
      const code = node.children[0];
      if (!code || code.type !== "element" || code.tagName !== "code") return;
      if (hasChartLanguage(code.properties.className)) return;
      const title = code.properties.dataCodeBlockTitle;
      delete code.properties.dataCodeBlockTitle;
      const hasCopyButton = code.properties.dataCodeCopyButton === true || code.properties.dataCodeCopyButton === "true" || code.properties.dataCodeCopyButton === "";
      delete code.properties.dataCodeCopyButton;
      if (typeof title !== "string" && !hasCopyButton) return;
      const children = [];
      if (typeof title === "string") {
        children.push({
          type: "element",
          tagName: "div",
          properties: { className: ["code-block-title"] },
          children: [{ type: "text", value: title }]
        });
      }
      const wrapperClassName = hasCopyButton ? [WRAPPER_CLASS, COPYABLE_CLASS] : [WRAPPER_CLASS];
      if (hasCopyButton) {
        copyTargetCount += 1;
        const copyTargetId = `code-copy-target-${String(copyTargetCount)}`;
        code.properties.dataCodeCopyTarget = copyTargetId;
        children.push({
          type: "element",
          tagName: "button",
          properties: {
            type: "button",
            className: [COPY_BUTTON_CLASS],
            dataCodeCopy: "",
            dataCodeCopyTarget: copyTargetId
          },
          children: [{ type: "text", value: "Copy" }]
        });
      }
      children.push(node);
      const wrapped = {
        type: "element",
        tagName: "div",
        properties: { className: wrapperClassName },
        children
      };
      parent.children[index] = wrapped;
    });
  };
};
var rehypeCodeBlockTitles_default = rehypeCodeBlockTitles;
var CHART_LANGUAGE_PREFIX2 = "language-chart-";
var CODE_LINE_CLASS = "code-line";
var CODE_LINE_HIGHLIGHTED_CLASS = "code-line-highlighted";
var CODE_LINE_NUMBERED_CLASS = "code-line-numbered";
var CODE_LINE_NUMBER_CLASS = "code-line-number";
function hasChartLanguage2(className) {
  const values = Array.isArray(className) ? className : typeof className === "string" ? [className] : [];
  return values.some((value) => typeof value === "string" && value.startsWith(CHART_LANGUAGE_PREFIX2));
}
function getHighlightLines(value) {
  const lines = /* @__PURE__ */ new Set();
  if (Array.isArray(value)) {
    for (const item of value) {
      if (typeof item === "number" && Number.isInteger(item) && item > 0) {
        lines.add(item);
      }
      if (typeof item === "string" && /^\d+$/.test(item)) {
        lines.add(Number(item));
      }
    }
    return lines;
  }
  if (typeof value === "string") {
    for (const fragment of value.split(/[^\d]+/)) {
      const trimmed = fragment.trim();
      if (/^\d+$/.test(trimmed)) {
        lines.add(Number(trimmed));
      }
    }
  }
  return lines;
}
function getShowLineNumbers(value) {
  return value === true || value === "true" || value === "";
}
function ensureClassNameList(className) {
  if (Array.isArray(className)) {
    return className.filter((entry) => typeof entry === "string");
  }
  if (typeof className === "string") {
    return [className];
  }
  return [];
}
function splitCodeChildrenIntoLines(children) {
  const lines = [[]];
  for (const child of children) {
    if (child.type !== "text") {
      const current = lines[lines.length - 1];
      if (!current) {
        continue;
      }
      current.push(child);
      continue;
    }
    const segments = child.value.split("\n");
    for (let i = 0; i < segments.length; i += 1) {
      const segment = segments[i];
      if (segment && segment.length > 0) {
        const current = lines[lines.length - 1];
        if (!current) {
          continue;
        }
        current.push({ type: "text", value: segment });
      }
      if (i < segments.length - 1) {
        lines.push([]);
      }
    }
  }
  return lines;
}
function isEmptyLine(line) {
  return line.length === 0 || line.every((node) => node.type === "text" && node.value.length === 0);
}
var rehypeCodeLineHighlights = function rehypeCodeLineHighlights2() {
  return (tree) => {
    visit(tree, "element", (node) => {
      if (node.tagName !== "code") return;
      if (hasChartLanguage2(node.properties.className)) return;
      const highlightLines = getHighlightLines(node.properties.dataCodeLineHighlights);
      delete node.properties.dataCodeLineHighlights;
      const showLineNumbers = getShowLineNumbers(node.properties.dataCodeLineNumbers);
      delete node.properties.dataCodeLineNumbers;
      if (highlightLines.size === 0 && !showLineNumbers) return;
      if (showLineNumbers) {
        const classNames = ensureClassNameList(node.properties.className);
        if (!classNames.includes(CODE_LINE_NUMBERED_CLASS)) {
          classNames.push(CODE_LINE_NUMBERED_CLASS);
        }
        node.properties.className = classNames;
      }
      const splitLines = splitCodeChildrenIntoLines(node.children);
      const lastLine = splitLines[splitLines.length - 1];
      const hasTrailingNewline = splitLines.length > 0 && !!lastLine && isEmptyLine(lastLine);
      const contentLines = hasTrailingNewline ? splitLines.slice(0, -1) : splitLines;
      const children = [];
      for (let i = 0; i < contentLines.length; i += 1) {
        const lineNumber = i + 1;
        const className = highlightLines.has(lineNumber) ? [CODE_LINE_CLASS, CODE_LINE_HIGHLIGHTED_CLASS] : [CODE_LINE_CLASS];
        const lineChildren = [];
        if (showLineNumbers) {
          lineChildren.push({
            type: "element",
            tagName: "span",
            properties: { className: [CODE_LINE_NUMBER_CLASS] },
            children: [{ type: "text", value: String(lineNumber) }]
          });
        }
        lineChildren.push(...contentLines[i] ?? []);
        const lineSpan = {
          type: "element",
          tagName: "span",
          properties: { className },
          children: lineChildren
        };
        children.push(lineSpan);
        if (i < contentLines.length - 1 || hasTrailingNewline) {
          const newline = { type: "text", value: "\n" };
          children.push(newline);
        }
      }
      node.children = children;
    });
  };
};
var rehypeCodeLineHighlights_default = rehypeCodeLineHighlights;

// src/internal/chartTypes.ts
var CHART_BLOCK_TYPES = [
  "line",
  "bar",
  "radar",
  "doughnut",
  "pie",
  "polarArea",
  "bubble",
  "scatter"
];
var CHART_BLOCK_TYPE_SET = new Set(CHART_BLOCK_TYPES);

// src/internal/rehypeChartBlocks.ts
var LANG_PREFIX = "language-chart-";
function chartTypeFromClassName(className) {
  const list = Array.isArray(className) ? className : typeof className === "string" ? [className] : [];
  for (const c of list) {
    if (typeof c === "string" && c.startsWith(LANG_PREFIX)) {
      return c.slice(LANG_PREFIX.length);
    }
  }
  return void 0;
}
var rehypeChartBlocks = function rehypeChartBlocks2(options) {
  const emitDiagnostic = options?.emitDiagnostic;
  return (tree) => {
    visit(tree, "element", (node, index, parent) => {
      if (node.tagName !== "pre") return;
      if (!parent || typeof index !== "number") return;
      const code = node.children[0];
      if (!code || code.type !== "element" || code.tagName !== "code") return;
      const chartType = chartTypeFromClassName(code.properties.className);
      if (!chartType || !CHART_BLOCK_TYPE_SET.has(chartType)) return;
      const raw = toText(code, { whitespace: "pre" }).trim();
      try {
        const parsed = JSON.parse(raw);
        const jsonStr = JSON.stringify(parsed);
        const div = {
          type: "element",
          tagName: "div",
          properties: {
            className: ["chart-mount"],
            dataChart: chartType,
            dataChartData: jsonStr
          },
          children: []
        };
        parent.children[index] = div;
      } catch {
        emitDiagnostic?.({
          code: "chart.invalid_json",
          message: `Chart JSON could not be parsed for type "${chartType}".`,
          severity: "warning",
          source: "chart"
        });
      }
    });
  };
};
var rehypeChartBlocks_default = rehypeChartBlocks;
var REMOTE_SRC_PATTERN = /^(https?:\/\/|data:|blob:)/;
function getClassList(className) {
  if (Array.isArray(className)) {
    return className.filter((value) => typeof value === "string");
  }
  if (typeof className === "string") {
    return [className];
  }
  return [];
}
function ensureInlineClass(node) {
  const classes = getClassList(node.properties.className);
  if (!classes.includes("markdown-inline-img")) {
    classes.push("markdown-inline-img");
  }
  node.properties.className = classes;
}
function rehypeCdnImages(options) {
  const kind = options?.kind;
  const slug = options?.slug;
  const map = options?.manifest?.map;
  const remote = options?.manifest?.remote;
  const emitDiagnostic = options?.emitDiagnostic;
  return (tree) => {
    visit(tree, "element", (node) => {
      if (node.tagName !== "img") return;
      ensureInlineClass(node);
      if (!kind || !slug) return;
      const src = node.properties.src;
      if (typeof src !== "string") return;
      if (REMOTE_SRC_PATTERN.test(src)) {
        const nextSrc2 = remote?.[src]?.url;
        if (nextSrc2) node.properties.src = nextSrc2;
        return;
      }
      if (src.startsWith("/")) return;
      const cleanPath = src.replace(/^\.\/+/, "");
      const key = `${kind}/${slug}/${cleanPath}`;
      const nextSrc = map?.[key]?.url;
      if (nextSrc) {
        node.properties.src = nextSrc;
      } else {
        emitDiagnostic?.({
          code: "images.manifest_miss",
          message: `No manifest entry found for "${key}".`,
          severity: "warning",
          source: "images"
        });
      }
    });
  };
}
var TRAILING_ID = /\s+\{#([\w-]+)\}\s*$/;
function remarkHeadingIds() {
  return function(tree) {
    visit(tree, "heading", (node) => {
      const last = node.children[node.children.length - 1];
      if (!last || last.type !== "text") return;
      const match = TRAILING_ID.exec(last.value);
      if (!match?.[1]) return;
      if (!node.data) node.data = {};
      if (!node.data.hProperties) node.data.hProperties = {};
      const id = match[1];
      const data = node.data;
      data.id = id;
      data.hProperties.id = id;
      data.hProperties.dataHeadingCustomId = "true";
      last.value = last.value.slice(0, match.index).replace(/\s+$/, "");
    });
  };
}
var SUPPORTED_CALLOUT_TYPES = [
  "note",
  "tip",
  "warning",
  "info",
  "success",
  "danger",
  "error",
  "caution",
  "important"
];
var DEFAULT_TITLES = {
  note: "Note",
  tip: "Tip",
  warning: "Warning",
  info: "Info",
  success: "Success",
  danger: "Danger",
  error: "Error",
  caution: "Caution",
  important: "Important"
};
function isSupportedCalloutType(value) {
  return SUPPORTED_CALLOUT_TYPES.includes(value);
}
function createMarkerParagraph(value) {
  return {
    type: "paragraph",
    children: [{ type: "text", value }]
  };
}
function normalizeWhitespace(value) {
  return value.replace(/\s+/g, " ").trim();
}
function createTitleNode(title) {
  return {
    type: "paragraph",
    data: {
      hName: "div",
      hProperties: {
        className: ["callout-title"]
      }
    },
    children: [{ type: "text", value: title }]
  };
}
function toPlainMarkdownNodes(name, label, children) {
  const openMarker = label ? `:::${name}[${label}]` : `:::${name}`;
  return [createMarkerParagraph(openMarker), ...children, createMarkerParagraph(":::")];
}
function extractDirectiveLabel(children) {
  const first = children[0];
  if (!first || first.type !== "paragraph") {
    return { label: null, bodyChildren: children };
  }
  const hasDirectiveLabelFlag = first.data?.directiveLabel === true;
  if (!hasDirectiveLabelFlag) {
    return { label: null, bodyChildren: children };
  }
  const label = first.children.map((child) => child.type === "text" ? child.value : "").join("");
  const normalizedLabel = normalizeWhitespace(label);
  return {
    label: normalizedLabel || null,
    bodyChildren: children.slice(1)
  };
}
function remarkCallouts() {
  return (tree) => {
    visit(tree, "containerDirective", (node, index, parent) => {
      if (index === void 0 || !parent) {
        return;
      }
      const directive = node;
      const name = directive.name;
      const { label, bodyChildren } = extractDirectiveLabel(directive.children);
      if (isSupportedCalloutType(name)) {
        const title = label ?? DEFAULT_TITLES[name];
        directive.children = [createTitleNode(title), ...bodyChildren];
        directive.data = {
          ...directive.data ?? {},
          hName: "div",
          hProperties: {
            ...directive.data?.hProperties ?? {},
            className: ["callout", `callout-${name}`]
          }
        };
        return;
      }
      const replacement = toPlainMarkdownNodes(name, label, bodyChildren);
      parent.children.splice(index, 1, ...replacement);
      return [SKIP, index + replacement.length];
    });
  };
}
var CHART_LANGUAGE_PREFIX3 = "chart-";
var TITLE_META_PATTERN = /(?:^|\s)title="([^"]*)"(?:\s|$)/;
var LINE_HIGHLIGHT_PATTERN = /\{([^}]*)\}/;
var LINE_NUMBERS_PATTERN = /(?:^|\s)(?:showLineNumbers|lineNumbers)(?:\s|$)/;
var COPY_BUTTON_PATTERN = /(?:^|\s)(?:showCopyButton|copyButton)(?:\s|$)/;
function getCodeBlockTitle(meta) {
  if (typeof meta !== "string") {
    return void 0;
  }
  const match = TITLE_META_PATTERN.exec(meta);
  if (!match) {
    return void 0;
  }
  return match[1];
}
function getCodeBlockHighlightLines(meta) {
  if (typeof meta !== "string") {
    return [];
  }
  const match = LINE_HIGHLIGHT_PATTERN.exec(meta);
  const raw = match?.[1];
  if (!raw) {
    return [];
  }
  const highlights = /* @__PURE__ */ new Set();
  const fragments = raw.split(",");
  for (const fragment of fragments) {
    const value = fragment.trim();
    if (/^\d+$/.test(value)) {
      const line = Number(value);
      if (line > 0) {
        highlights.add(line);
      }
      continue;
    }
    const rangeMatch = /^(\d+)-(\d+)$/.exec(value);
    if (!rangeMatch) {
      continue;
    }
    const start = Number(rangeMatch[1]);
    const end = Number(rangeMatch[2]);
    if (start <= 0 || end <= 0 || start > end) {
      continue;
    }
    for (let line = start; line <= end; line += 1) {
      highlights.add(line);
    }
  }
  return Array.from(highlights).sort((a, b) => a - b);
}
function hasCodeBlockLineNumbers(meta) {
  return typeof meta === "string" && LINE_NUMBERS_PATTERN.test(meta);
}
function hasCodeBlockCopyButton(meta) {
  return typeof meta === "string" && COPY_BUTTON_PATTERN.test(meta);
}
var remarkCodeBlockTitles = function remarkCodeBlockTitles2() {
  return (tree) => {
    visit(tree, "code", (node) => {
      if (typeof node.lang === "string" && node.lang.startsWith(CHART_LANGUAGE_PREFIX3)) {
        return;
      }
      const title = getCodeBlockTitle(node.meta);
      const highlights = getCodeBlockHighlightLines(node.meta);
      const hasLineNumbers = hasCodeBlockLineNumbers(node.meta);
      const hasCopyButton = hasCodeBlockCopyButton(node.meta);
      if (title === void 0 && highlights.length === 0 && !hasLineNumbers && !hasCopyButton) {
        return;
      }
      node.data ??= {};
      node.data.hProperties ??= {};
      if (title !== void 0) {
        node.data.hProperties.dataCodeBlockTitle = title;
      }
      if (highlights.length > 0) {
        node.data.hProperties.dataCodeLineHighlights = highlights;
      }
      if (hasLineNumbers) {
        node.data.hProperties.dataCodeLineNumbers = true;
      }
      if (hasCopyButton) {
        node.data.hProperties.dataCodeCopyButton = true;
      }
    });
  };
};
var baseAttrs = defaultSchema.attributes ?? {};
var baseStar = baseAttrs["*"] ?? [];
var baseSpan = baseAttrs.span ?? [];
var baseDiv = baseAttrs.div ?? [];
var baseCode = baseAttrs.code ?? [];
var baseButton = baseAttrs.button ?? [];
var baseH1 = baseAttrs.h1 ?? [];
var baseH2 = baseAttrs.h2 ?? [];
var baseH3 = baseAttrs.h3 ?? [];
var baseH4 = baseAttrs.h4 ?? [];
var baseH5 = baseAttrs.h5 ?? [];
var baseH6 = baseAttrs.h6 ?? [];
var baseTags = defaultSchema.tagNames ?? [];
var calloutClassNames = [
  "callout",
  "callout-note",
  "callout-tip",
  "callout-warning",
  "callout-info",
  "callout-success",
  "callout-danger",
  "callout-error",
  "callout-caution",
  "callout-important",
  "callout-title",
  "chart-mount",
  "code-block",
  "code-block-title",
  "code-block-copyable"
];
var buttonClassNames = ["code-block-copy-button"];
var markdownSanitizeSchema = {
  ...defaultSchema,
  tagNames: [...baseTags, "button"],
  attributes: {
    ...baseAttrs,
    "*": [...baseStar, "className"],
    span: [...baseSpan, "style"],
    code: [...baseCode, "className", "dataCodeCopyTarget", "dataCodeLineHighlights", "dataCodeLineNumbers"],
    button: [...baseButton, ["className", ...buttonClassNames], "type", "dataCodeCopy", "dataCodeCopyTarget"],
    h1: [...baseH1, "dataHeadingCustomId"],
    h2: [...baseH2, "dataHeadingCustomId"],
    h3: [...baseH3, "dataHeadingCustomId"],
    h4: [...baseH4, "dataHeadingCustomId"],
    h5: [...baseH5, "dataHeadingCustomId"],
    h6: [...baseH6, "dataHeadingCustomId"],
    div: [
      ...baseDiv,
      ["className", ...calloutClassNames],
      ["dataChart", ...CHART_BLOCK_TYPES],
      "dataChartData"
    ]
  }
};
var svgTags = [
  "svg",
  "g",
  "path",
  "marker",
  "rect",
  "circle",
  "line",
  "polygon",
  "text",
  "tspan",
  "style",
  "foreignObject",
  "defs",
  "symbol"
];
var svgSharedAttributes = [
  "style",
  "transform",
  "fill",
  "stroke",
  "strokeWidth",
  "strokeDasharray",
  "id",
  "className"
];
var schemaWithSvg = {
  ...markdownSanitizeSchema,
  tagNames: [...markdownSanitizeSchema.tagNames ?? [], ...svgTags],
  attributes: {
    ...markdownSanitizeSchema.attributes ?? {},
    div: [
      ...baseDiv,
      "style",
      "xmlns",
      ["className", ...calloutClassNames],
      ["dataChart", ...CHART_BLOCK_TYPES],
      "dataChartData"
    ],
    svg: [
      ...svgSharedAttributes,
      "ariaRoleDescription",
      "role",
      "viewBox",
      "width",
      "xmlns"
    ],
    g: [
      ...svgSharedAttributes,
      "dataId",
      "dataNode"
    ],
    path: [...svgSharedAttributes, "d", "markerEnd"],
    marker: [
      ...svgSharedAttributes,
      "markerHeight",
      "markerUnits",
      "markerWidth",
      "orient",
      "refX",
      "refY",
      "viewBox"
    ],
    rect: [...svgSharedAttributes, "height", "rx", "ry", "width", "x", "y", "name"],
    circle: [...svgSharedAttributes, "cx", "cy", "r"],
    line: [...svgSharedAttributes, "markerEnd", "x1", "x2", "y1", "y2"],
    polygon: [...svgSharedAttributes, "points"],
    text: [
      ...svgSharedAttributes,
      "alignmentBaseline",
      "dominantBaseline",
      "dy",
      "textAnchor",
      "x",
      "y"
    ],
    tspan: [...svgSharedAttributes, "dy", "x"],
    foreignObject: ["height", "width"],
    symbol: ["clipRule", "fillRule", "height", "id", "width"],
    defs: [],
    style: []
  }
};
function rehypeStripHeadingCustomIdMarker() {
  return (tree) => {
    visit(tree, "element", (node) => {
      if (!/^h[1-6]$/.test(node.tagName)) {
        return;
      }
      delete node.properties.dataHeadingCustomId;
    });
  };
}
function createMarkdownProcessorInternal(options) {
  const gates = options?.features ?? {
    math: true,
    code: true,
    codeHighlighting: true,
    charts: true,
    sanitize: true
  };
  const processor = unified().use(remarkParse).use(remarkGfm).use(remarkDirective);
  if (gates.math) {
    processor.use(remarkMath);
  }
  processor.use(remarkHeadingIds).use(remarkCallouts);
  if (gates.code) {
    processor.use(remarkCodeBlockTitles);
  }
  processor.use(remarkRehype, { allowDangerousHtml: true }).use(rehypeRaw);
  if (gates.charts) {
    if (options?.emitDiagnostic) {
      processor.use(rehypeChartBlocks_default, { emitDiagnostic: options.emitDiagnostic });
    } else {
      processor.use(rehypeChartBlocks_default);
    }
  }
  if (gates.code) {
    processor.use(rehypeCodeBlockTitles_default);
  }
  if (gates.math) {
    processor.use(rehypeKatex, { output: "html" });
  }
  if (gates.sanitize) {
    processor.use(rehypeSanitize, markdownSanitizeSchema);
  }
  processor.use(rehypeSlug).use(rehypeAutolinkHeadings, { behavior: "wrap" });
  if (gates.code && gates.codeHighlighting) {
    processor.use(rehypeHighlight).use(rehypeCodeLineHighlights_default);
  }
  if (!options?.preserveHeadingCustomIdMarker) {
    processor.use(rehypeStripHeadingCustomIdMarker);
  }
  processor.use(rehypeCdnImages, options).use(rehypeStringify);
  return processor;
}

export { createMarkdownProcessorInternal, rehypeCdnImages, rehypeChartBlocks_default, rehypeCodeBlockTitles_default, rehypeCodeLineHighlights_default, remarkCallouts, remarkCodeBlockTitles, remarkHeadingIds, schemaWithSvg };
