// featherdown/dist/chunk-CWUDY7YI.js
import { createMarkdownProcessorInternal } from './chunk-YKLQHPRB.js';
import yaml from 'js-yaml';
import { toText } from 'hast-util-to-text';
import { visit, EXIT } from 'unist-util-visit';

// src/featherdownDiagnosticsError.ts
var FeatherdownDiagnosticsError = class extends Error {
  name = "FeatherdownDiagnosticsError";
  diagnostics;
  /**
   * Full parse result as built before the strict check, including `html` and
   * the same `diagnostics` array as {@link FeatherdownDiagnosticsError.diagnostics}.
   * Omitted when the error is constructed without a result (internal use only).
   */
  result;
  constructor(diagnostics, result) {
    super(`Featherdown strict diagnostics failed with ${String(diagnostics.length)} diagnostic(s).`);
    this.diagnostics = diagnostics;
    this.result = result;
    Object.setPrototypeOf(this, new.target.prototype);
  }
};

// src/internal/buildFeatherdownResult.ts
function buildStyleAssets(resolved) {
  const styles = ["featherdown/styles/base.css"];
  if (resolved.math.enabled) {
    styles.push("featherdown/styles/katex.css");
  }
  if (resolved.code.enabled) {
    styles.push("featherdown/styles/code.css");
    if (resolved.code.highlighting) {
      styles.push("featherdown/styles/highlight.css");
    }
  }
  return styles;
}
function buildAssets(resolved, mermaidFeature) {
  return {
    styles: buildStyleAssets(resolved),
    scripts: [],
    features: {
      math: resolved.math.enabled,
      code: resolved.code.enabled,
      charts: resolved.charts,
      mermaid: mermaidFeature
    }
  };
}
function buildFeatherdownResult(args) {
  const { body, frontmatter, metadata, document: doc, options: resolved, mermaidFeature } = args;
  const mermaidEnabled = mermaidFeature === true;
  let diagnostics = doc.diagnostics;
  if (!resolved.diagnostics.enabled) {
    diagnostics = [];
  }
  let toc = doc.toc;
  let headings = doc.headings;
  if (!resolved.headings.enabled) {
    toc = [];
    headings = [];
  } else if (!resolved.headings.toc) {
    toc = [];
  }
  let excerpt = doc.excerpt;
  let wordCount = doc.wordCount;
  let estimatedReadingMinutes = doc.estimatedReadingMinutes;
  if (!resolved.publishing.excerpt) {
    excerpt = null;
  }
  if (!resolved.publishing.wordCount) {
    wordCount = 0;
  }
  if (!resolved.publishing.readingTime) {
    estimatedReadingMinutes = 0;
  }
  return {
    html: doc.html,
    diagnostics,
    toc,
    headings,
    excerpt,
    wordCount,
    estimatedReadingMinutes,
    body,
    frontmatter,
    metadata,
    stats: {
      wordCount,
      readingTimeMinutes: estimatedReadingMinutes
    },
    assets: buildAssets(resolved, mermaidEnabled)
  };
}

// src/internal/buildMetadataFromFrontmatter.ts
function buildMetadataFromFrontmatter(frontmatter) {
  const metadata = {};
  if (typeof frontmatter.title === "string") {
    metadata.title = frontmatter.title;
  }
  if (typeof frontmatter.description === "string") {
    metadata.description = frontmatter.description;
  }
  if (typeof frontmatter.date === "string") {
    metadata.date = frontmatter.date;
  }
  if (typeof frontmatter.updated === "string") {
    metadata.updated = frontmatter.updated;
  }
  if (typeof frontmatter.status === "string") {
    metadata.status = frontmatter.status;
  }
  if (typeof frontmatter.author === "string") {
    metadata.author = frontmatter.author;
  }
  const tags = frontmatter.tags;
  if (Array.isArray(tags) && tags.length > 0 && tags.every((t) => typeof t === "string")) {
    metadata.tags = tags.filter((t) => typeof t === "string");
  }
  return metadata;
}

// src/internal/featherdownParseHelpers.ts
function pickRenderMarkdownOptions(merged) {
  if (!merged) {
    return void 0;
  }
  const { kind, slug, manifest } = merged;
  if (kind === void 0 && slug === void 0 && manifest === void 0) {
    return void 0;
  }
  const r = {};
  if (kind !== void 0) {
    r.kind = kind;
  }
  if (slug !== void 0) {
    r.slug = slug;
  }
  if (manifest !== void 0) {
    r.manifest = manifest;
  }
  return r;
}
function toProcessorGates(resolved) {
  return {
    math: resolved.math.enabled,
    code: resolved.code.enabled,
    codeHighlighting: resolved.code.enabled && resolved.code.highlighting,
    charts: resolved.charts,
    sanitize: resolved.sanitize
  };
}
var FRONT_MATTER_FENCE = /^---\r?\n([\s\S]*?)\r?\n---(?:\r?\n|$)/;
function extractFrontMatter(raw, yamlMode) {
  if (!raw.startsWith("---")) {
    return { body: raw, frontmatter: {}, hasFrontmatter: false };
  }
  const match = FRONT_MATTER_FENCE.exec(raw);
  if (!match) {
    return { body: raw, frontmatter: {}, hasFrontmatter: false };
  }
  const yamlSource = match[1] ?? "";
  const bodyAfterFence = raw.slice(match[0].length);
  try {
    const loaded = yaml.load(yamlSource);
    if (loaded === void 0 || loaded === null) {
      return { body: bodyAfterFence, frontmatter: {}, hasFrontmatter: true };
    }
    if (typeof loaded !== "object" || Array.isArray(loaded)) {
      if (yamlMode === "strict") {
        throw new Error(
          "Front matter root must be a YAML mapping (object), not a list or scalar."
        );
      }
      return { body: raw, frontmatter: {}, hasFrontmatter: false };
    }
    return {
      body: bodyAfterFence,
      frontmatter: loaded,
      hasFrontmatter: true
    };
  } catch (e) {
    if (yamlMode === "strict") {
      if (e instanceof Error && e.message.startsWith("Front matter")) {
        throw e;
      }
      const detail = e instanceof Error ? e.message : String(e);
      throw new Error(`Front matter YAML could not be parsed: ${detail}`);
    }
    return { body: raw, frontmatter: {}, hasFrontmatter: false };
  }
}
function extractFrontMatterStrict(raw) {
  return extractFrontMatter(raw, "strict");
}
function extractFrontMatterLenient(raw) {
  return extractFrontMatter(raw, "lenient");
}
function parseFrontmatter(markdown, mode) {
  if (mode === false) {
    return { body: markdown, frontmatter: {}, hasFrontmatter: false };
  }
  return extractFrontMatterLenient(markdown);
}

// src/internal/diagnostics.ts
function createDiagnosticCollector() {
  const diagnostics = [];
  const emit = (d) => {
    diagnostics.push(d);
  };
  return {
    emit,
    diagnostics
  };
}

// src/internal/renderMarkdownDocumentInternal.ts
var WORDS_PER_MINUTE = 200;
var excerptTagNames = /* @__PURE__ */ new Set(["p", "blockquote", "li"]);
function normalizeText(text) {
  return text.replace(/\s+/g, " ").trim();
}
function getHeadingDepth(tagName) {
  const match = /^h([1-6])$/.exec(tagName);
  if (!match) {
    return null;
  }
  return Number(match[1]);
}
function isParagraphLikeTagName(tagName) {
  return excerptTagNames.has(tagName);
}
function isCustomHeadingId(value) {
  return value === true || value === "true" || value === "";
}
function extractHeadings(root) {
  const headings = [];
  visit(root, "element", (node) => {
    const depth = getHeadingDepth(node.tagName);
    if (!depth) {
      return;
    }
    const id = typeof node.properties.id === "string" ? node.properties.id : "";
    if (!id) {
      return;
    }
    const text = normalizeText(toText(node));
    if (!text) {
      return;
    }
    headings.push({
      index: headings.length,
      depth,
      text,
      id,
      hasCustomId: isCustomHeadingId(node.properties.dataHeadingCustomId)
    });
    delete node.properties.dataHeadingCustomId;
  });
  return headings;
}
function extractToc(headings) {
  return headings.map((heading) => ({
    depth: heading.depth,
    text: heading.text,
    id: heading.id
  }));
}
function extractExcerpt(root) {
  let excerpt = null;
  visit(root, "element", (node) => {
    if (excerpt !== null) {
      return EXIT;
    }
    if (!isParagraphLikeTagName(node.tagName)) {
      return;
    }
    const text = normalizeText(toText(node));
    if (!text) {
      return;
    }
    excerpt = text;
    return EXIT;
  });
  return excerpt;
}
function countWords(root) {
  const text = normalizeText(toText(root));
  if (!text) {
    return 0;
  }
  const words = text.match(/\S+/g);
  return words ? words.length : 0;
}
function estimateReadingMinutes(wordCount) {
  if (wordCount <= 0) {
    return 0;
  }
  return Math.max(1, Math.ceil(wordCount / WORDS_PER_MINUTE));
}
async function renderMarkdownDocumentInternal(markdown, options, pipelineFeatures) {
  const collector = createDiagnosticCollector();
  const processor = createMarkdownProcessorInternal({
    ...options,
    emitDiagnostic: collector.emit,
    preserveHeadingCustomIdMarker: true,
    ...pipelineFeatures !== void 0 ? { features: pipelineFeatures } : {}
  });
  const parsed = processor.parse(markdown);
  const root = await processor.run(parsed);
  const headings = extractHeadings(root);
  const html = String(processor.stringify(root));
  const wordCount = countWords(root);
  return {
    html,
    diagnostics: collector.diagnostics,
    toc: extractToc(headings),
    headings,
    excerpt: extractExcerpt(root),
    wordCount,
    estimatedReadingMinutes: estimateReadingMinutes(wordCount)
  };
}

// src/internal/resolveFeatherdownOptions.ts
var DEFAULT_MATH = { enabled: true };
var DEFAULT_CODE = {
  enabled: true,
  highlighting: true,
  lineNumbers: false,
  copyButton: true
};
var DEFAULT_HEADINGS = {
  enabled: true,
  slugs: true,
  anchors: true,
  toc: true
};
var DEFAULT_PUBLISHING = {
  excerpt: true,
  wordCount: true,
  readingTime: true
};
function mergeFeatherdownMathGroup(ctor, parse) {
  if (parse !== void 0) {
    if (typeof parse === "boolean") {
      return parse;
    }
    if (ctor !== void 0 && typeof ctor === "object") {
      return { ...ctor, ...parse };
    }
    return { ...parse };
  }
  return ctor;
}
function mergeFeatherdownCodeGroup(ctor, parse) {
  if (parse !== void 0) {
    if (typeof parse === "boolean") {
      return parse;
    }
    if (ctor !== void 0 && typeof ctor === "object") {
      return { ...ctor, ...parse };
    }
    return { ...parse };
  }
  return ctor;
}
function mergeFeatherdownHeadingsGroup(ctor, parse) {
  if (parse !== void 0) {
    if (typeof parse === "boolean") {
      return parse;
    }
    if (ctor !== void 0 && typeof ctor === "object") {
      return { ...ctor, ...parse };
    }
    return { ...parse };
  }
  return ctor;
}
function mergeFeatherdownPublishingGroup(ctor, parse) {
  if (parse !== void 0) {
    if (typeof parse === "boolean") {
      return parse;
    }
    if (ctor !== void 0 && typeof ctor === "object") {
      return { ...ctor, ...parse };
    }
    return { ...parse };
  }
  return ctor;
}
function mergeFeatherdownOptions(constructorOptions, parseOptions) {
  if (constructorOptions === void 0 && parseOptions === void 0) {
    return void 0;
  }
  const out = { ...constructorOptions, ...parseOptions };
  const mathMerged = mergeFeatherdownMathGroup(constructorOptions?.math, parseOptions?.math);
  if (mathMerged !== void 0) {
    out.math = mathMerged;
  }
  const codeMerged = mergeFeatherdownCodeGroup(constructorOptions?.code, parseOptions?.code);
  if (codeMerged !== void 0) {
    out.code = codeMerged;
  }
  const headingsMerged = mergeFeatherdownHeadingsGroup(constructorOptions?.headings, parseOptions?.headings);
  if (headingsMerged !== void 0) {
    out.headings = headingsMerged;
  }
  const publishingMerged = mergeFeatherdownPublishingGroup(constructorOptions?.publishing, parseOptions?.publishing);
  if (publishingMerged !== void 0) {
    out.publishing = publishingMerged;
  }
  return out;
}
function normalizeMath(value, base) {
  if (value === void 0) {
    return { ...base };
  }
  if (typeof value === "boolean") {
    return { enabled: value };
  }
  return {
    enabled: value.enabled ?? base.enabled
  };
}
function normalizeCode(value, base) {
  if (value === void 0) {
    return { ...base };
  }
  if (typeof value === "boolean") {
    return {
      ...base,
      enabled: value
    };
  }
  return {
    enabled: value.enabled ?? base.enabled,
    highlighting: value.highlighting ?? base.highlighting,
    lineNumbers: value.lineNumbers ?? base.lineNumbers,
    copyButton: value.copyButton ?? base.copyButton
  };
}
function normalizeHeadings(value, base) {
  if (value === void 0) {
    return { ...base };
  }
  if (typeof value === "boolean") {
    if (!value) {
      return {
        enabled: false,
        slugs: false,
        anchors: false,
        toc: false
      };
    }
    return { ...DEFAULT_HEADINGS };
  }
  return {
    enabled: true,
    slugs: value.slugs ?? base.slugs,
    anchors: value.anchors ?? base.anchors,
    toc: value.toc ?? base.toc
  };
}
function normalizePublishing(value, base) {
  if (value === void 0) {
    return { ...base };
  }
  if (typeof value === "boolean") {
    if (!value) {
      return { excerpt: false, wordCount: false, readingTime: false };
    }
    return { ...DEFAULT_PUBLISHING };
  }
  return {
    excerpt: value.excerpt ?? base.excerpt,
    wordCount: value.wordCount ?? base.wordCount,
    readingTime: value.readingTime ?? base.readingTime
  };
}
function normalizeFrontmatter(value) {
  if (value === true || value === "auto") {
    return value;
  }
  return false;
}
function normalizeDiagnostics(value) {
  if (value === void 0) {
    return { enabled: true, mode: "warn" };
  }
  if (value === false) {
    return { enabled: false, mode: "warn" };
  }
  if (value === true) {
    return { enabled: true, mode: "warn" };
  }
  return { enabled: true, mode: value };
}
function resolveFeatherdownOptions(raw) {
  const mergedNested = raw ?? {};
  return {
    ...mergedNested.kind !== void 0 ? { kind: mergedNested.kind } : {},
    ...mergedNested.slug !== void 0 ? { slug: mergedNested.slug } : {},
    ...mergedNested.manifest !== void 0 ? { manifest: mergedNested.manifest } : {},
    math: normalizeMath(mergedNested.math, DEFAULT_MATH),
    code: normalizeCode(mergedNested.code, DEFAULT_CODE),
    sanitize: mergedNested.sanitize ?? true,
    diagnostics: normalizeDiagnostics(mergedNested.diagnostics),
    charts: mergedNested.charts ?? true,
    headings: normalizeHeadings(mergedNested.headings, DEFAULT_HEADINGS),
    publishing: normalizePublishing(mergedNested.publishing, DEFAULT_PUBLISHING),
    frontmatter: normalizeFrontmatter(mergedNested.frontmatter)
  };
}

// src/internal/throwIfStrictDiagnosticsViolated.ts
function throwIfStrictDiagnosticsViolated(resolved, result) {
  if (resolved.diagnostics.mode !== "strict" || result.diagnostics.length === 0) {
    return;
  }
  throw new FeatherdownDiagnosticsError(result.diagnostics, result);
}

// src/featherdown.ts
var Featherdown = class {
  defaultOptions;
  constructor(options) {
    this.defaultOptions = options;
  }
  /**
   * Render Markdown to HTML and document metadata. Constructor options are merged
   * with per-call options; per-call values win. Nested option groups use a
   * shallow one-level merge.
   */
  async parse(markdown, parseOptions) {
    const mergedOptions = mergeFeatherdownOptions(this.defaultOptions, parseOptions);
    const resolvedOptions = resolveFeatherdownOptions(mergedOptions);
    const legacy = pickRenderMarkdownOptions(mergedOptions);
    const fm = parseFrontmatter(markdown, resolvedOptions.frontmatter);
    const metadata = buildMetadataFromFrontmatter(fm.frontmatter);
    const document = await renderMarkdownDocumentInternal(fm.body, legacy, toProcessorGates(resolvedOptions));
    const result = buildFeatherdownResult({
      body: fm.body,
      frontmatter: fm.frontmatter,
      metadata,
      document,
      options: resolvedOptions
    });
    throwIfStrictDiagnosticsViolated(resolvedOptions, result);
    return result;
  }
};

export { Featherdown, FeatherdownDiagnosticsError, buildFeatherdownResult, buildMetadataFromFrontmatter, createDiagnosticCollector, extractFrontMatterStrict, mergeFeatherdownOptions, parseFrontmatter, pickRenderMarkdownOptions, renderMarkdownDocumentInternal, resolveFeatherdownOptions, throwIfStrictDiagnosticsViolated, toProcessorGates };
