// featherdown/dist/index.js
export { createMarkdownProcessor, rehypeCdnImages, rehypeChartBlocks } from './chunk-DNLCIOR6.js';
import { renderMarkdownDocumentInternal, extractFrontMatterStrict } from './chunk-CWUDY7YI.js';
export { Featherdown, FeatherdownDiagnosticsError } from './chunk-CWUDY7YI.js';
import './chunk-YKLQHPRB.js';

// src/libraryId.ts
function libraryId() {
  return "featherdown";
}

// src/renderMarkdownDocument.ts
async function renderMarkdownDocument(markdown, options) {
  return renderMarkdownDocumentInternal(markdown, options);
}

// src/renderMarkdown.ts
async function renderMarkdown(markdown, options) {
  const result = await renderMarkdownDocument(markdown, options);
  return { html: result.html, diagnostics: result.diagnostics };
}

// src/renderMarkdownToHtml.ts
async function renderMarkdownToHtml(markdown, options) {
  const result = await renderMarkdownDocument(markdown, options);
  return result.html;
}

// src/parseMarkdownFile.ts
function parseMarkdownFile(raw) {
  const { body, frontmatter } = extractFrontMatterStrict(raw);
  return { frontMatter: frontmatter, content: body };
}

export { libraryId, parseMarkdownFile, renderMarkdown, renderMarkdownDocument, renderMarkdownToHtml };
