// featherdown/dist/node.js
import { Featherdown, mergeFeatherdownOptions, resolveFeatherdownOptions, pickRenderMarkdownOptions, parseFrontmatter, buildMetadataFromFrontmatter, toProcessorGates, buildFeatherdownResult, throwIfStrictDiagnosticsViolated, createDiagnosticCollector } from './chunk-CWUDY7YI.js';
export { FeatherdownDiagnosticsError } from './chunk-CWUDY7YI.js';
import { remarkHeadingIds, remarkCallouts, remarkCodeBlockTitles, rehypeChartBlocks_default, rehypeCodeBlockTitles_default, schemaWithSvg, rehypeCodeLineHighlights_default, rehypeCdnImages } from './chunk-YKLQHPRB.js';
import rehypeAutolinkHeadings from 'rehype-autolink-headings';
import rehypeHighlight from 'rehype-highlight';
import rehypeKatex from 'rehype-katex';
import rehypeMermaid from 'rehype-mermaid';
import rehypeRaw from 'rehype-raw';
import rehypeSanitize from 'rehype-sanitize';
import rehypeSlug from 'rehype-slug';
import rehypeStringify from 'rehype-stringify';
import remarkDirective from 'remark-directive';
import remarkGfm from 'remark-gfm';
import remarkMath from 'remark-math';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import { unified } from 'unified';
import { visit, EXIT } from 'unist-util-visit';
import { readFile } from 'fs/promises';
import { toText } from 'hast-util-to-text';

function rehypeStripHeadingCustomIdMarker() {
  return (tree) => {
    visit(tree, "element", (node) => {
      if (!/^h[1-6]$/.test(node.tagName)) {
        return;
      }
      delete node.properties.dataHeadingCustomId;
    });
  };
}
function createMarkdownProcessorWithMermaidInternal(options) {
  const gates = options?.features ?? {
    math: true,
    code: true,
    codeHighlighting: true,
    charts: true,
    sanitize: true
  };
  const fallbackToOriginalCodeBlock = (element) => element;
  const processor = unified().use(remarkParse).use(remarkGfm).use(remarkDirective);
  if (gates.math) {
    processor.use(remarkMath);
  }
  processor.use(remarkHeadingIds).use(remarkCallouts);
  if (gates.code) {
    processor.use(remarkCodeBlockTitles);
  }
  processor.use(remarkRehype, { allowDangerousHtml: true }).use(rehypeRaw);
  if (gates.charts) {
    if (options?.emitDiagnostic) {
      processor.use(rehypeChartBlocks_default, { emitDiagnostic: options.emitDiagnostic });
    } else {
      processor.use(rehypeChartBlocks_default);
    }
  }
  processor.use(rehypeSlug).use(rehypeAutolinkHeadings, { behavior: "wrap" });
  if (gates.code) {
    processor.use(rehypeCodeBlockTitles_default);
  }
  if (gates.math) {
    processor.use(rehypeKatex, { output: "html" });
  }
  if (gates.code && gates.codeHighlighting) {
    processor.use(rehypeHighlight);
  }
  processor.use(rehypeMermaid, {
    strategy: "inline-svg",
    errorFallback: fallbackToOriginalCodeBlock
  });
  if (gates.sanitize) {
    processor.use(rehypeSanitize, schemaWithSvg);
  }
  if (gates.code && gates.codeHighlighting) {
    processor.use(rehypeCodeLineHighlights_default);
  }
  if (!options?.preserveHeadingCustomIdMarker) {
    processor.use(rehypeStripHeadingCustomIdMarker);
  }
  processor.use(rehypeCdnImages, options).use(rehypeStringify);
  return processor;
}
var WORDS_PER_MINUTE = 200;
var excerptTagNames = /* @__PURE__ */ new Set(["p", "blockquote", "li"]);
function normalizeText(text) {
  return text.replace(/\s+/g, " ").trim();
}
function getHeadingDepth(tagName) {
  const match = /^h([1-6])$/.exec(tagName);
  if (!match) {
    return null;
  }
  return Number(match[1]);
}
function isParagraphLikeTagName(tagName) {
  return excerptTagNames.has(tagName);
}
function isCustomHeadingId(value) {
  return value === true || value === "true" || value === "";
}
function extractHeadings(root) {
  const headings = [];
  visit(root, "element", (node) => {
    const depth = getHeadingDepth(node.tagName);
    if (!depth) {
      return;
    }
    const id = typeof node.properties.id === "string" ? node.properties.id : "";
    if (!id) {
      return;
    }
    const text = normalizeText(toText(node));
    if (!text) {
      return;
    }
    headings.push({
      index: headings.length,
      depth,
      text,
      id,
      hasCustomId: isCustomHeadingId(node.properties.dataHeadingCustomId)
    });
    delete node.properties.dataHeadingCustomId;
  });
  return headings;
}
function extractToc(headings) {
  return headings.map((heading) => ({
    depth: heading.depth,
    text: heading.text,
    id: heading.id
  }));
}
function extractExcerpt(root) {
  let excerpt = null;
  visit(root, "element", (node) => {
    if (excerpt !== null) {
      return EXIT;
    }
    if (!isParagraphLikeTagName(node.tagName)) {
      return;
    }
    const text = normalizeText(toText(node));
    if (!text) {
      return;
    }
    excerpt = text;
    return EXIT;
  });
  return excerpt;
}
function countWords(root) {
  const text = normalizeText(toText(root));
  if (!text) {
    return 0;
  }
  const words = text.match(/\S+/g);
  return words ? words.length : 0;
}
function estimateReadingMinutes(wordCount) {
  if (wordCount <= 0) {
    return 0;
  }
  return Math.max(1, Math.ceil(wordCount / WORDS_PER_MINUTE));
}
async function renderMarkdownDocumentWithMermaidInternal(markdown, options, pipelineFeatures) {
  const collector = createDiagnosticCollector();
  const processor = createMarkdownProcessorWithMermaidInternal({
    ...options,
    emitDiagnostic: collector.emit,
    preserveHeadingCustomIdMarker: true,
    ...pipelineFeatures !== void 0 ? { features: pipelineFeatures } : {}
  });
  const parsed = processor.parse(markdown);
  const root = await processor.run(parsed);
  const headings = extractHeadings(root);
  const html = String(processor.stringify(root));
  const wordCount = countWords(root);
  return {
    html,
    diagnostics: collector.diagnostics,
    toc: extractToc(headings),
    headings,
    excerpt: extractExcerpt(root),
    wordCount,
    estimatedReadingMinutes: estimateReadingMinutes(wordCount)
  };
}

// src/nodeFeatherdown.ts
function stripMermaidForBase(options) {
  if (!options) {
    return void 0;
  }
  const rest = { ...options };
  delete rest.mermaid;
  return Object.keys(rest).length > 0 ? rest : void 0;
}
function mergeNodeFeatherdownOptions(ctor, parse) {
  const baseMerged = mergeFeatherdownOptions(stripMermaidForBase(ctor), stripMermaidForBase(parse));
  const mermaid = parse?.mermaid !== void 0 ? parse.mermaid : ctor?.mermaid !== void 0 ? ctor.mermaid : void 0;
  if (mermaid === void 0) {
    return baseMerged;
  }
  return { ...baseMerged ?? {}, mermaid };
}
function isMermaidSvgEnabled(merged) {
  return merged?.mermaid !== void 0 && merged.mermaid !== false;
}
var Featherdown2 = class {
  base;
  ctorOptions;
  constructor(options) {
    this.base = new Featherdown(stripMermaidForBase(options));
    this.ctorOptions = options;
  }
  async parse(markdown, parseOptions) {
    const mergedNode = mergeNodeFeatherdownOptions(this.ctorOptions, parseOptions);
    if (!isMermaidSvgEnabled(mergedNode)) {
      return this.base.parse(markdown, stripMermaidForBase(mergedNode));
    }
    const mergedOptions = mergeFeatherdownOptions(stripMermaidForBase(this.ctorOptions), stripMermaidForBase(parseOptions));
    const resolvedOptions = resolveFeatherdownOptions(mergedOptions);
    const legacy = pickRenderMarkdownOptions(mergedOptions);
    const fm = parseFrontmatter(markdown, resolvedOptions.frontmatter);
    const metadata = buildMetadataFromFrontmatter(fm.frontmatter);
    const document = await renderMarkdownDocumentWithMermaidInternal(
      fm.body,
      legacy,
      toProcessorGates(resolvedOptions)
    );
    const result = buildFeatherdownResult({
      body: fm.body,
      frontmatter: fm.frontmatter,
      metadata,
      document,
      options: resolvedOptions,
      mermaidFeature: true
    });
    throwIfStrictDiagnosticsViolated(resolvedOptions, result);
    return result;
  }
  async parseFile(path, parseOptions) {
    const markdown = await readFile(path, "utf8");
    return this.parse(markdown, parseOptions);
  }
};

// src/node.ts
async function renderMarkdownToHtmlWithMermaid(markdown, options) {
  const processor = createMarkdownProcessorWithMermaidInternal(options);
  const file = await processor.process(markdown);
  return String(file);
}

export { Featherdown2 as Featherdown, renderMarkdownToHtmlWithMermaid };
